aten/src/ATen/native/hammerblade/Offload.h

```diff
@@ -23,10 +23,71 @@ typedef struct {
+  args.push_back(arg0);                                                                                 \
+  std::vector<Scalar> scalars;                                                                          \
+  offload_tensor_scalar_impl(args, scalars, kernel);                                                    \
+} while (0);
+
+#define HB_OFFLOAD_TENSOR_KERNEL_3ARGS(arg0, arg1, kernel)                                              \
+do {                                                                                                    \
+  std::vector<Tensor> args;                                                                             \
+  args.push_back(arg0);                                                                                 \
+  args.push_back(arg1);                                                                                 \
+  std::vector<Scalar> scalars;                                                                          \
+  offload_tensor_scalar_impl(args, scalars, kernel);                                                    \
+} while (0);
+
+#define HB_OFFLOAD_TENSOR_KERNEL_4ARGS(arg0, arg1, arg2, kernel)                                        \
+do {                                                                                                    \
+  std::vector<Tensor> args;                                                                             \
+  args.push_back(arg0);                                                                                 \
+  args.push_back(arg1);                                                                                 \
+  args.push_back(arg2);                                                                                 \
+  std::vector<Scalar> scalars;                                                                          \
+  offload_tensor_scalar_impl(args, scalars, kernel);                                                    \
+} while (0);
+
+#define HB_OFFLOAD_TENSOR_KERNEL_5ARGS(arg0, arg1, arg2, arg3, kernel)                                  \
+do {                                                                                                    \
+  std::vector<Tensor> args;                                                                             \
+  args.push_back(arg0);                                                                                 \
+  args.push_back(arg1);                                                                                 \
+  args.push_back(arg2);                                                                                 \
+  args.push_back(arg3);                                                                                 \
+  std::vector<Scalar> scalars;                                                                          \
+  offload_tensor_scalar_impl(args, scalars, kernel);                                                    \
```

Comment:
I'm missing something here... how do we populate scalar arguments?

Reply:
These marcos are for the kernels that use tensor arguments only, and thus scalars are empty.
I was trying to use the same offloading function (`offload_tensor_scalar_impl`), which handles arbitrary number of tensors followed by  arbitrary number of scalars.

Reply:
I think we might need to start splitting the `Offload.h/cpp` into multiple files. It is growing fast.

Reply:
Yes, I felt that too when adding conv offloading. 

I believe we can have `Offload.h` include multiple offload related headers so that we could do `#include Offload.h` to access all offloading routines -- like we do now.

Reply:
> These marcos are for the kernels that use tensor arguments only, and thus scalars are empty.
> I was trying to use the same offloading function (`offload_tensor_scalar_impl`), which handles arbitrary number of tensors followed by arbitrary number of scalars.

Variadic arguments?

Reply:
I think we could do 

```
template <typename... T1, typename T2>
offload_tensor_scalar_impl(T1... tensors, T2... scalars, char* kernel) {
}
```

We don't need to template types here... but when I tried this, I was seeing compilation error without templated types.

Reply:
ah good point. Let's merge this one and your conv2d in, and then let's think about what should be our offloading interface. I guess we need to make some decisions so we won't keep changing things